azurelinuxagent/ga/exthandlers.py

```diff
@@ -294,53 +293,43 @@ def __last_gs_unsupported(self):
                self.__gs_aggregate_status.code == GoalStateAggregateStatusCodes.GoalStateUnsupportedRequiredFeatures
 
     def run(self):
-        etag, activity_id, correlation_id, gs_creation_time = None, None, None, None
-
         try:
-            extensions_goal_state = self.protocol.get_goal_state().extensions_goal_state
+            gs = self.protocol.get_goal_state()
+            egs = gs.extensions_goal_state
 
-            # self.ext_handlers and etag need to be initialized first, since status reporting depends on them; also
+            # self.ext_handlers needs to be initialized before returning, since status reporting depends on it; also
             # we make a deep copy of the extensions, since changes are made to self.ext_handlers while processing the extensions
-            self.ext_handlers = copy.deepcopy(extensions_goal_state.extensions)
-            etag = self.protocol.client.get_goal_state().incarnation
+            self.ext_handlers = copy.deepcopy(egs.extensions)
 
             if not self._extension_processing_allowed():
                 return
 
-            gs_creation_time = extensions_goal_state.created_on_timestamp
-            activity_id = extensions_goal_state.activity_id
-            correlation_id = extensions_goal_state.correlation_id
+            utc_start = datetime.datetime.utcnow()
+            error = None
+            message = "ProcessExtensionsGoalState started [{0} channel: {1} source: {2} activity: {3} correlation {4} created: {5}]".format(
+                egs.id, egs.channel, egs.source, egs.activity_id, egs.correlation_id, egs.created_on_timestamp)
+            logger.info(message)
+            add_event(op=WALAEventOperation.ExtensionProcessing, message=message)
+
+            try:
+                self.__process_and_handle_extensions(gs.incarnation)  # TODO: review the use of incarnation
+                self._cleanup_outdated_handlers()
+            except Exception as e:
+                error = u"Error processing extensions:{0}".format(textutil.format_exception(e))
+            finally:
+                duration = elapsed_milliseconds(utc_start)
+                if error is None:
+                    message = 'ProcessExtensionsGoalState completed [{0} {1} ms]'.format(egs.id, duration)
+                    logger.info(message)
+                else:
+                    message = 'ProcessExtensionsGoalState failed [{0} {1} ms]\n{2}'.format(egs.id, duration, error)
+                    logger.error(message)
+                add_event(op=WALAEventOperation.ExtensionProcessing, is_success=(error is None), message=message, log_event=False, duration=duration)
+
         except Exception as error:
```

Comment:
Why move this down and nest the above try except? I'm not quite seeing what this covers over where it was before.

Reply:
i find a single try/except more readable than 2 separate blocks. the new implementation preserves the previous behavior